### ISSUE TITLE
Errors are silent, but hugely impacting

### ISSUE THREAD

**Issue by NightJar** [silverstripe/silverstripe-graphql#181]
I have recently run into a use case where mid-development of a feature I made a pull from the upstream repository.
Unbeknownst to me at the time some GraphQL fields had been added to a query, however being that this was defined in `yaml` things were destined not to work until a `flush` had taken place.

As I was not aware I simply loaded the CMS without a flush... and nothing happened.

Literally nothing. Half of it loaded (HTML templates), but everything depending on a GraphQL query as the source of data just hung with a loading screen. There was no hint of an error besides a great lack of data. My console was quiet of errors, every XHR response was `200 OK`, and there were no UI updates to notify me of an issue.

Luckily a colleague informed me of the alteration, so I was quickly back on track, but if he had not been available at the time I would have been left to laboriously run through each XHR and check the output matched my expectations manually.

To this I see a few different approaches to solving this problem:

**HTTP response codes**
As I had asked the server for something it didn't know about, the problem lies with me and my request. I feel like there should be something in the 4xx range (like `422`) that is a far more appropriate response.

**UI Update**
However with rendering a 4xx response this resolves with most handlers as an error.

This could stop our UI from updating in the bad case... an argument against HTTP error codes - in which I make a counter case we should still update it to let the user (developer - I'm in `dev` mode) know.

In a good case we could still render what data we did receive (if any) with a caveat that some is potentially missing.

In any case with the `admin` module I think there's still plenty of space to render a global toast style notification that an issue has happened with our data request no matter the UI or section that is open at the time.

**tl;dr** there needs to be at least _something_ to let a developer know that things have gone awry.
Something more than a hung screen and a message buried deep within an "everything's fine" response.

![image](https://user-images.githubusercontent.com/778003/44966302-8700e600-af8e-11e8-9933-42fc0a17763a.png)

**Comment by ScopeyNZ** [comment]
It would be also pretty nice just to see any errors returned from the `graphql` endpoint show in the console. Maybe depending on the type of error? Something like "This graphql endpoint has no idea what you're talking about when you ask for this specific field` would be only a problem for a dev, as the JS app defines the GraphQL queries (usually).

**Comment by unclecheese** [comment]
I'm not sure this is an issue with the GraphQL module, though...?

We need to be better at GraphQL error handling across the board, but it should be an implementation of the UI, and I don't see it as a concern of this module. The GraphQL module provides all the data a developer needs to create a nice failure state in the UI. We do fairly [graceful error handling in asset-admin](https://github.com/silverstripe/silverstripe-asset-admin/blob/1/client/src/containers/Gallery/Gallery.js#L842), for instance.

With regard to HTTP status codes, I think of it like form validation. The request _is_ successful (i.e. you provided the required `query` parameter, and the method is GET/POST), but the business logic that deals with the value of that parameter rejected it. If I enter an invalid credit card number into a form, I get back a 200, because the request is valid, but my business logic has rejected the user input. At the end of the day, unless you're using [persisted queries](https://github.com/silverstripe/silverstripe-graphql/pull/179), GraphQL queries are user input, and as such subject to validation.

You also have issues around partial success, where, for instance, permissions might fail for one record out of 10.

If it's about error UI, maybe this issue needs to be moved to the module that lacks a nice error experience?

**Comment by tractorcow** [comment]
I've discovered this with graphql as well. The GraphQL Controller does respond to errors but even when rendering errors the HTTP code is 200. Wouldn't you agree that the standard web convention is to represent the status of the response via the status code, and that we should defer to the convention in the absence of good justification otherwise?

These responses are intended to be read by code not humans, so I would make it a responsibility of your front end to translate these http response code into human errors.

**Comment by tractorcow** [comment]
Maybe it could be an option? (use_http_codes) on controller?

**Comment by tractorcow** [comment]
I did a bit of digging and I found that errors are actually resolved to json responses within the webonyx module itself.

https://github.com/webonyx/graphql-php/issues/320

**Comment by ScopeyNZ** [comment]
I was talking about this with @unclecheese a little while ago. The lines get pretty blurry between the "RESTful" `HTTP 422 Unprocessable Entity` style response ("You gave me something bad, I can't handle it") vs. GraphQL saying "Hey, I understand you want this, here's where you're wrong, but I was fine, so it's 200 OK"

It's not too surprising that the webonyx module chose that. Perhaps we should provide some default middleware on the JS side to log errors to console when in dev mode?

**Comment by tractorcow** [comment]
ok, I think I've found a way.

In Manager::formatError() it calls `$exception->getPrevious()`. Maybe I can use that to throw the error directly there instead of rendering it. :)

The use case we have in mind is 426 errors when a user needs to re authenticate within a react application. It's not simply logging, but is a required part of the application behaviour.

**Comment by tractorcow** [comment]
Custom manager class:

```php
    public static function formatError(Error $exception)
    {
        // Force exceptions to be caught by HTTPController
        $previous = $exception->getPrevious();
        if ($previous instanceof Exception) {
            throw $previous;
        }

        return parent::formatError($exception);
    }
```

Custom Controller class:

```php
    } catch (Exception $exception) {
            // Use inner response if available
            if ($exception instanceof HTTPResponse_Exception) {
                $response->setStatusCode($exception->getCode());
            } else {
                $response->setStatusCode(500);
            }

            // Generate body
            $error = ['message' => $exception->getMessage()];
            if (Director::isDev()) {
                $error['code'] = $exception->getCode();
                $error['file'] = $exception->getFile();
                $error['line'] = $exception->getLine();
                // Note: getTrace() is huge so omit here
            }
            $result = [
                'errors' => [$error]
            ];
        }
```

**Comment by tractorcow** [comment]
ping @unclecheese has this been addressed in graphql 4?
